src/Encamina.Enmarcha.SemanticKernel/MemoryStoreExtender.cs

```diff
@@ -0,0 +1,165 @@
+
+            OnMemoryStore(new() { EventType = MemoryStoreEventTypes.Delete, MemoryId = memoryId, CollectionName = collectionName });
+        }
+    }
+
+    /// <inheritdoc cref="IMemoryStoreExtender.GetMemoryAsync(string, string, CancellationToken)" />
+    public virtual async Task<MemoryContent> GetMemoryAsync(string memoryId, string collectionName, CancellationToken cancellationToken)
+    {
+        var chunkSize = await GetChunkSize(memoryId, collectionName, cancellationToken);
+
+        if (chunkSize == 0)
+        {
+            return null;
+        }
+
+        var memoryRecords = await MemoryStore.GetBatchAsync(collectionName, Enumerable.Range(0, chunkSize)
+            .Select(i => BuildMemoryIdentifier(memoryId, i)), cancellationToken: cancellationToken)
+            .ToListAsync(cancellationToken);
+
+        return new MemoryContent
+        {
+            Metadata = JsonSerializer.Deserialize<Dictionary<string, string>>(memoryRecords[0].Metadata.AdditionalMetadata),
+            Chunks = memoryRecords.Select(m => m.Metadata.Text),
+        };
+    }
+
+    /// <inheritdoc cref="IMemoryStoreExtender.BatchUpsertMemoriesAsync(string, IDictionary{string, MemoryContent}, Kernel, CancellationToken)" />
+    public virtual async IAsyncEnumerable<string> BatchUpsertMemoriesAsync(string collectionName, IDictionary<string, MemoryContent> memoryContents, Kernel kernel, [EnumeratorCancellation] CancellationToken cancellationToken)
+    {
+        var memoryRecords = new Collection<MemoryRecord>();
+        var memoryIds = new List<string>();
+
+        foreach (var memoryContentItem in memoryContents)
+        {
+            var memoryContentId = memoryContentItem.Key;
+            var memoryContent = memoryContentItem.Value;
+            var totalChunks = memoryContent.Chunks.Count();
+
+            if (totalChunks > 0)
+            {
+                memoryContent.Metadata.Add(Constants.MetadataTotalChunksCount, totalChunks.ToString());
+
+                for (var i = 0; i < totalChunks; i++)
+                {
+                    var chunk = memoryContent.Chunks.ElementAt(i);
+                    var embedding = await kernel.GetRequiredService<ITextEmbeddingGenerationService>().GenerateEmbeddingAsync(chunk, kernel, cancellationToken);
+                    memoryRecords.Add(MemoryRecord.LocalRecord($@"{memoryContentId}-{i}", chunk, null, embedding, JsonSerializer.Serialize(memoryContent.Metadata)));
+                }
+            }
+        }
+
+        var memoryRecordsUniqueIdentifiers = MemoryStore.UpsertBatchAsync(collectionName, memoryRecords, cancellationToken);
+
+        foreach (var memoryId in memoryContents.Select(content => content.Key))
+        {
+            OnMemoryStore(new() { EventType = MemoryStoreEventTypes.Upsert, MemoryId = memoryId, CollectionName = collectionName });
```

Comment:
This think this might destroy the memory stack of any process if the batch _upsert_ has too many records 😱

Perhaps, it would be better to get all the memory Ids in a collection and then raise an event if type `UpsertBatch` with those IDs. This means that maybe the `MemoryId` property on `MemoryStoreEventArgs` should be a collection of strings rather than a single value. 

Finally, the variable `memoryRecordsUniqueIdentifiers` already has the memory IDs that the `foreach` loop was processing.

Reply:
Thank you for your contribution. You are right. I will update it.